### PR TITLE
Introduces ConstGet<T>

### PR DESCRIPTION
**Pull request by gupnik**
Issue: https://github.com/paritytech/substrate/issues/9967

### CODE REVIEW

**Review by shawntabrizi — APPROVED** [review]

File: bounded-collections/src/lib.rs [hunk 1]
```diff
@@ -41,15 +41,33 @@ pub trait Get<T> {
...
-impl<T: Default> Get<T> for () {
+/// A trait for querying a single value from a type.
+///
+/// The value should be a constant.
+pub trait ConstGet<T> {
+	/// Return the constant value.
+	fn get() -> T;
+}
+
+/// Implement Get for any type that implements ConstGet.
```
Comment (kianenigma) [inline bounded-collections/src/lib.rs:52]:
```suggestion
// Implement Get for any type that implements ConstGet.
```

impl blocks don't have rust-docs I think and should be documented with just `//`.

Comment (bkchr) [reply bounded-collections/src/lib.rs:52]: They can have docs. They also appear in the docs.

File: bounded-collections/src/lib.rs [hunk 2]
```diff
@@ -41,15 +41,33 @@ pub trait Get<T> {
...
+impl<T, U> Get<U> for T
```
Comment (KiChjang) [inline bounded-collections/src/lib.rs:53]: This implementation can be omitted if we just make `Get<T>` a supertrait of `ConstGet<T>`.

File: bounded-collections/src/lib.rs (hunk 1 above)
Comment (gupnik) [reply bounded-collections/src/lib.rs]: Used supertrait as suggested by @KiChjang.

File: bounded-collections/src/lib.rs (hunk 2 above)
Comment (gupnik) [reply bounded-collections/src/lib.rs]: Done, thanks!

File: bounded-collections/src/lib.rs [hunk 3]
```diff
@@ -76,6 +97,16 @@ macro_rules! impl_const_get {
...
 			}
 		}
+		impl<const T: $t> ConstGet<$t> for $name<T> {
+			fn get() -> $t {
+				T
+			}
+		}
+		impl<const T: $t> ConstGet<Option<$t>> for $name<T> {
+			fn get() -> Option<$t> {
+				Some(T)
```
Comment (KiChjang) [inline bounded-collections/src/lib.rs:107]: Why do we need this implementation?

Comment (gupnik) [reply bounded-collections/src/lib.rs:107]: It actually implements `ConstGet` for all `const` types.

Comment (KiChjang) [reply bounded-collections/src/lib.rs:107]: I meant why do we need an implementation of `ConstGet<Option<T>>` on the `Const*` types.

Comment (gupnik) [reply bounded-collections/src/lib.rs:107]: Ah, I was following a similar paradigm as `Get<Option<T>>` that has already been implemented for const types.

**Review by KiChjang — APPROVED** [review]

File: bounded-collections/src/lib.rs [hunk 4]
```diff
@@ -41,20 +41,41 @@ pub trait Get<T> {
...
 }
 
-/// Implement Get by returning Default for any type that implements Default.
+impl<T: Default> ConstGet<T> for () {
+	fn get() -> T {
+		T::default()
+	}
+}
+
+/// Implement ConstGet by returning Default for any type that implements Default.
```
Comment (ggwpez) [inline bounded-collections/src/lib.rs:64]: This comment looks out of place here.

File: bounded-collections/src/lib.rs [hunk 5]
```diff
@@ -41,20 +41,41 @@ pub trait Get<T> {
 	fn get() -> T;
 }
 
+/// A trait for querying a single value from a type.
+///
+/// The value should be a constant.
```
Comment (ggwpez) [inline bounded-collections/src/lib.rs:46]: Do i under stand correctly that the only difference to the normal `Get` trait is that this **MUST** be implemented by a const?  
I think that should be highlighted a bit better as to how it differs from `Get`.

**Review by ggwpez — COMMENTED** [review]
Test?

**Comment by gupnik** [comment]
@ggwpez I think @kianenigma mentioned that this doesn't add a lot of value as `Const` still doesn't enforce that the provided value is indeed a const and is more of a syntatic sugar at this point. So, will close this one.

**Pull request closed without merging**